### PR TITLE
Add new nextcloud chart

### PR DESCRIPTION
**Pull request by affix**
**What this PR does / why we need it**:
This PR adds a nextcloud chart to incubator 

Nextcloud offers industry-leading on-premises file sync and online collaboration technology. Our expertise is in combining the convenience and ease of use of consumer-grade solutions like Dropbox and Google Drive with the security, privacy and control business needs.

**Special notes for your reviewer**:
This is loosely based around the owncloud chart, However I as many other prefer to use nextcloud for some of its optimisations.

Uses the official Nextcloud Docker image

### CODE REVIEW

**Comment by bbriggs** [comment]
I was literally writing a chart for that this weekend. Glad to see someone else took an interest too.

File: incubator/nextcloud/Chart.yaml
Comment (mattfarina) [inline incubator/nextcloud/Chart.yaml]: The name should be your github username. That helps us to know who to contact when there's an issue with the chart.

File: incubator/nextcloud/OWNERS
Comment (mattfarina) [inline incubator/nextcloud/OWNERS]: For the time you can drop this file. It looks like you copied this from a chart managed by Bitnami. Since you're not a member of the Kubernetes org at the moment this isn't going to enable you to review or merge for the chart. Once you have some more contributions you can apply for membership and then a file like this will make sense.

File: incubator/nextcloud/values.yaml
Comment (mattfarina) [inline incubator/nextcloud/values.yaml]: > It is generally not advisable to provide hard resource limits to workloads and leave it configurable unless the workload requires such quantity bare minimum to function.

This is from the r[eviewers guide](https://github.com/kubernetes/charts/blob/master/REVIEW_GUIDELINES.md#kubernetes-native-workloads).

Can you comment these out unless this is the minimum to function.

File: incubator/nextcloud/values.yaml
Comment (mattfarina) [inline incubator/nextcloud/values.yaml]: Does the container image allow for both http and https? It might be nice to allow for secure out of the box. That doesn't need to be part of the first PR but would be nice to add.

File: incubator/nextcloud/values.yaml
Comment (mattfarina) [inline incubator/nextcloud/values.yaml]: Can you update the image information to follow our standard layout? See [the guide](https://github.com/kubernetes/charts/blob/master/REVIEW_GUIDELINES.md#configuration) for more information and an example.

Also, please use a tagged version rather than latest. This will enable consistency between environments.

File: incubator/nextcloud/templates/ingress.yaml
Comment (mattfarina) [inline incubator/nextcloud/templates/ingress.yaml]: Why is the API version a values parameter?

File: incubator/nextcloud/templates/deployment.yaml
Comment (mattfarina) [inline incubator/nextcloud/templates/deployment.yaml]: Can you use the `apps/v1beta2` version instead?

File: incubator/nextcloud/requirements.yaml
Comment (mattfarina) [inline incubator/nextcloud/requirements.yaml]: The version supports ranges. What about `~0.7.0` so that it can pull in patch level differences?

**Review by mattfarina — COMMENTED** [review]
Thanks for submitting the pull request. I know of several people who want to use nextcloud on Kubernetes. This is great.

I did a quick review to provide some feedback.

**Comment by unguiculus** [comment]
Please compare with what you get from `helm create` using Helm 2.8+ and update accordingly.

**Comment by pierreozoux** [comment]
I think it would make more sense to PR against `stable` folder. It is a pretty straight forward addition.

Also I can help maintain this chart.

Another question: how could we keep this under `/nextcloud` github namespace, and sync it with the official charts? This would help a lot the pace of development.

**Comment by billimek** [comment]
@affix are you still planning to finish this?  I was about to write a chart for nextcloud as well, but will wait for this to complete - unless you want someone else to pick up where you left off?

**Comment by affix** [comment]
@billimek I am planning to, just been a hectic few weeks. I will get it sorted and address comments

**Comment by affix** [comment]
@mattfarina I have made the comments you requested

File: stable/nextcloud/README.md
Comment (unguiculus) [inline stable/nextcloud/README.md]: Should be 1.9+

File: stable/nextcloud/templates/deployment.yaml
Comment (unguiculus) [inline stable/nextcloud/templates/deployment.yaml]: apps/v1

File: stable/nextcloud/.helmignore
Comment (unguiculus) [inline stable/nextcloud/.helmignore]: Use `.helmignore` you get from `helm create`

File: stable/nextcloud/values.yaml
Comment (unguiculus) [inline stable/nextcloud/values.yaml:36]: I'd suggest you add one level of nesting:

```yaml
nextcloud:
  host:
  loadBalancerIP:
  username:
  password:
  ...
```

**Review by unguiculus — CHANGES_REQUESTED** [review]
Again, please compare with what you get from helm create using Helm 2.8+ and update accordingly (`_helpers.tpl`, standard labels, etc.).

Also, you'll need to merge in the latest master to make CI happy.

**Comment by billimek** [comment]
The [following](https://github.com/affix/charts/blob/nextcloud/stable/nextcloud/templates/deployment.yaml#L36-L40) also needs to change in order for the nextcloud autoconfiguration to work properly,

```
        - name: nextcloud_DATABASE_NAME
          value: {{ .Values.mariadb.mariadbDatabase | quote }}
        - name: nextcloud_DATABASE_USER
          value: {{ .Values.mariadb.mariadbUser | quote }}
        - name: nextcloud_DATABASE_PASSWORD
```

should be:

```
        - name: MYSQL_DATABASE
          value: {{ .Values.mariadb.mariadbDatabase | quote }}
        - name: MYSQL_USER
          value: {{ .Values.mariadb.mariadbUser | quote }}
        - name: MYSQL_PASSWORD
```

**Comment by billimek** [comment]
https://github.com/affix/charts/pull/1 created with the changed I suggested.

Also added the ability to set existing claims for use in NextCloud via the 'external storage' plugin.

**Comment by pierreozoux** [comment]
What do we need to move this further?

**Comment by mattfarina** [comment]
@pierreozoux There is open feedback that needs to be addressed on the PR.

**Comment by affix** [comment]
@pierreozoux @mattfarina I'll get this fixed within the next week, been a bit busy with other things lately

**Comment by affix** [comment]
thanks for the contribution @billimek I think I've made the required changes to the chart

**Comment by unteem** [comment]
Hi @affix. I think the apache-data volume is not necessary. The volumes should be: `/var/www/data` `/var/www/config` `/var/www/apps` and `/var/www/custom_apps`

File: stable/nextcloud/README.md
Comment (tilosp) [inline stable/nextcloud/README.md:3]: this should be https://nextcloud.com/

Comment (mattfarina) [reply stable/nextcloud/README.md:3]: Good catch. https://nextcloud.org gives a certificate error.

File: stable/nextcloud/values.yaml
Comment (tujian) [inline stable/nextcloud/values.yaml:45]: This ref [https://github.com/bitnami/bitnami-docker-nextcloud#environment-variables](https://github.com/bitnami/bitnami-docker-nextcloud#environment-variables), which is substituted from [https://github.com/bitnami/bitnami-docker-owncloud#environment-variables](https://github.com/bitnami/bitnami-docker-owncloud#environment-variables), leads to 404.

**Comment by suvash** [comment]
Tried templating and running this today. It required some additional changes. The patch is fairly large, so I've included it as a gist on https://gist.github.com/suvash/d4cc3b2a767945014e29fb69bc9dd712

I can verify that it works on GKE after the following changes.
- with mariadb
  - without persistent volume
  - with persistent volume

**Comment by suvash** [comment]
@affix Figured out it was easier to open a PR on your repo instead.

File: stable/nextcloud/Chart.yaml
Comment (giacope) [inline stable/nextcloud/Chart.yaml:2]: Why is the chart version `0.1.1`? Shouldn't it be  `0.1.0` since it's a new chart?

**Comment by billimek** [comment]
I’ve never seen a good rule for this. When there is a PR with multiple iterations of requested changes, why do we not expect to see the minor version bumped multiple times?  

When testing locally it is extremely difficult to know which changes you are picking up if you artificially keep the version pegged to a static value.

**Comment by giacomoguiulfo** [comment]
@billimek AFAIK, commits are squashed when merged so the different versions will be lost or not exist from the master branch, just the last one present will be there. Therefore, it can be possible to have a `0.1.0` and a `0.1.2` but NOT a `0.1.1`, which doesn't make too mush sense IMO.

**Comment by billimek** [comment]
> @billimek AFAIK, commits are squashed when merged so the different versions will be lost or not exist from the master branch, just the last one present will be there. Therefore, it can be possible to have a `0.1.0` and a `0.1.2` but NOT a `0.1.1`, which doesn't make too mush sense IMO.

Thanks @giacomoguiulfo, that explanation makes sense.  Do you know of a way to iterate through local changes & testing of a chart (before pushing to an active PR) in a way that you can be confident that you are testing a new set of changes while keeping the chart version set to an unchanging static value?

e.g. chart version is pegged to 0.1.0 but you go through 6 iterations of changes locally, yet `helm search <chart name>` will show 0.1.0 for each of those 6 changes.  How can you be confident that you are testing a change?

**Comment by billimek** [comment]
Is there anything we are waiting on to merge this one?

File: stable/nextcloud/Chart.yaml
Comment (giacope) [inline stable/nextcloud/Chart.yaml:2]: So apparently, charts going to stable need to start at `1.0.0`

File: stable/nextcloud/README.md
Comment (anarcat) [inline stable/nextcloud/README.md:3]: making this formal with GH's new proposed change gizmo:

```suggestion
[nextcloud](https://nextcloud.com/) is a file sharing server that puts the control and security of your own data back into your hands.
```

File: stable/nextcloud/values.yaml
Comment (anarcat) [inline stable/nextcloud/values.yaml:2]:
```suggestion
## Official nextcloud image version 
## ref: https://hub.docker.com/r/library/nextcloud/tags/
```

File: stable/nextcloud/values.yaml
Comment (anarcat) [inline stable/nextcloud/values.yaml:45]:
```suggestion
## ref: https://github.com/nextcloud/docker#auto-configuration-via-environment-variables
```

File: stable/nextcloud/values.yaml
Comment (anarcat) [inline stable/nextcloud/values.yaml:75]: another broken link:

```suggestion
  ## ref: https://github.com/nextcloud/docker#using-an-external-database
```

File: stable/nextcloud/values.yaml
Comment (anarcat) [inline stable/nextcloud/values.yaml:80]: just remove this duplicate link

```suggestion
```

File: stable/nextcloud/values.yaml
Comment (anarcat) [inline stable/nextcloud/values.yaml:85]: just remove this duplicate link

```suggestion
```

File: stable/nextcloud/values.yaml
Comment (anarcat) [inline stable/nextcloud/values.yaml:90]: just remove this duplicate link

```suggestion
```

**Review by anarcat — CHANGES_REQUESTED** [review]
so the bitnami docker image is gone, as others have noted here. i've tried to fix all the 404s refering to that image and point instead at the official docker image managed by the nextcloud organization. i am not sure it's sufficient here because the newer image might not work the same way esp. WRT the mariadb bitnami image.

**Comment by varac** [comment]
Hej, what's the state of this ? Would be great to have it meged soon!

**Comment by iMartyn** [comment]
FYI the official nextcloud docker image lacks php-imap support because a maintainer deemed it to be "too big".  This makes it useless for my setup, even if the helm chart works. :'(

**Comment by anarcat** [comment]
On 2018-11-22 05:12:51, Martyn Ranyard wrote:
> FYI the official nextcloud docker image lacks php-imap support because a maintainer deemed it to be "too big".  This makes it useless for my setup, even if the helm chart works. :'(

Well we have to stat somewhere. There can be a billion variations for
any PHP-hosted project but hopefully the chart would serve as a good
foundation for any of those...

**Comment by joshuacox** [comment]
+1 on getting this merged

**Comment by doodlemania2** [comment]
Pretty please?

**Comment by kfox1111** [comment]
If we can get something merged, then it will be easier for people to contribute bugfixes/features.

**Comment by joshuacox** [comment]
perhaps @affix abandoned this PR?  What is left from @unguiculus review that is undone?  I agree with @kfox1111 let's wrap this up and get it merged to it can get some wider testing and then new PRs.

**Comment by ngortheone** [comment]
Also interested in this chart. I think it is reasonable to merge this to incubating so then interested people can contribute fixes

File: stable/nextcloud/Chart.yaml
Comment (desaintmartin) [reply stable/nextcloud/Chart.yaml:2]:
```suggestion
version: 1.0.0
```

**Review by desaintmartin — CHANGES_REQUESTED** [review]

**Comment by anarcat** [comment]
if someone is really interested in getting this off the ground, the way forward is to just make a new PR, because the proposed changes here do not seem to be going anywhere...

**Comment by EmperorArthur** [comment]
Came looking for a Nextcloud chart and found this PR.  I see a few things haven't been fixed from the changes requested.  I'd fix it, but am still learning.  Hope it's still being actively worked on.

PS:  The `Details` button on "tide" goes to a 404.

**Comment by chrisingenhaag** [comment]
Hey together, 
I created a new, bit refactored and updated the chart. PR is #10922 . I hope I´ve got all of your feedback and that we can get this chart into stable.

**Comment by evanstucker-hates-2fa** [comment]
I'm looking forward to this getting merged.

**Comment by unguiculus** [comment]
I'd suggest we move forward with #10922.

**Comment by unguiculus** [comment]
Closing in favor of #10922.

**Pull request closed without merging**
